RESOURCE_ACCESS_CONTROL_FOR_PLUGINS.md

```diff
@@ -0,0 +1,493 @@
+void doExecute(Task task, ShareResourceRequest request, ActionListener<ShareResourceResponse> listener) {
+    if (request.getResourceId() == null || request.getResourceId().isEmpty()) {
+        listener.onFailure(new IllegalArgumentException("Resource ID cannot be null or empty"));
+        return;
+    }
+
+    ResourceSharingClient resourceSharingClient = ResourceSharingClientAccessor.getResourceSharingClient(nodeClient, settings);
+    resourceSharingClient.shareResource(
+            request.getResourceId(),
+            RESOURCE_INDEX_NAME,
+            request.getShareWith(),
+            ActionListener.wrap(sharing -> {
+                ShareResourceResponse response = new ShareResourceResponse(sharing.getShareWith());
+                listener.onResponse(response);
+            }, listener::onFailure)
+    );
+}
+```
+
+
+---
+
+## **5. What are Scopes?**
+
+This feature introduces a new **sharing mechanism** called **scopes**. Scopes define **the level of access** granted to users for a resource. They are **defined and maintained by plugins**, and the security plugin does **not** interpret or enforce their specific meanings. This approach gives plugins the **flexibility** to define scope names and behaviors based on their use case.
+
+Each plugin must **document its scope definitions** so that users understand the **sharing semantics** and how different scopes affect access control.
+
+Scopes enable **granular access control**, allowing shareableResources to be shared with **customized permission levels**, making the system more flexible and adaptable to different use cases.
+
+### **Common Scopes for Plugins to declare**
+| Scope        | Description                                         |
+|-------------|-----------------------------------------------------|
+| `PUBLIC`    | The resource is accessible to all users.            |
+| `READ_ONLY` | Users can view but not modify the resource.         |
+| `READ_WRITE` | Users can view and modify the resource.             |
+
+By default, all shareableResources are private and only visible to the owner and super-admins. Resources become accessible to others only when explicitly shared.
+
+SPI provides you an interface, with two default scopes `PUBLIC` and `READ_ONLY`, which can be extended to introduce more plugin-specific values.
+
+Here’s a cleaned-up version of your last section for better clarity and structure:
+
+#### **Example: PUBLIC Scoped Document**
+If a resource is intended to be publicly accessible, it must be shared using the following structure:
+
+```json
+{
+    "share_with": {
+        "public": {
+            "backend_roles": ["*"],
+            "roles": ["*"],
+            "users": ["*"]
+        }
+    }
+}
+```
+
+In this structure:
+- `"backend_roles": ["*"]` → Grants access to all backend roles.
+- `"roles": ["*"]` → Grants access to all roles.
+- `"users": ["*"]` → Grants access to all users.
+
+**This ensures that the resource is accessible to everyone.**
```

Comment:
Could you elaborate what's here the difference to the following one:

```json
{
    "share_with": {
        "read_only": {
            "backend_roles": ["*"],
            "roles": ["*"],
            "users": ["*"]
        }
    }
}
```

Reply:
the names are only semantic atm. This will in future be mapped to action groups. The discussion can be found under this thread: https://github.com/opensearch-project/security/pull/5185#issuecomment-2744341082